prototype_source/maskedtensor_adagrad.rst

```diff
@@ -0,0 +1,182 @@
+Efficiency of writing "sparse" semantics for Adagrad
+====================================================
+
+`Issue 1369 <https://github.com/pytorch/pytorch/issues/1369>`__ discussed the additional lines of code
+that were introduce while writing "sparse" semantics for Adagrad.
+But really the code doesn't use sparsity as a compression and optimization technique,
+it wants to use masked semantics. We worked around this by introducing one-off semantics and operators
+that encode this behavior while forcing users to be aware of storage details such as indices and values.
+
+In particular we'll point out when sparsity is used as a semantic extension, i.e. unspecified values are not zero
+and when it is just used to compress zeros.
+We'll also compare and contrast this with equivalent code written using MaskedTensor.
+In the end the code snippets are repeat without additional comments to show the difference in brevity.
+
+Set up
+------
+
+.. code-block:: python
+
+    import torch
+    from torch.masked.maskedtensor import masked_tensor
+
+    # Some hyperparameters
+    eps = 1e-10
+    clr = 0.1
+
+    i = torch.tensor([[0, 1, 1], [2, 0, 2]])
+    v = torch.tensor([3, 4, 5], dtype=torch.float32)
+    grad = torch.sparse_coo_tensor(i, v, [2, 4])
+
+Original sparse implementation
+------------------------------
+
+First, let's break down the current implementation of 
+`Adagrad (functional) <https://github.com/pytorch/pytorch/blob/6c2f235d368b697072699e5ca9485fd97d0b9bcc/torch/optim/_functional.py#L16-L51>`__
+in PyTorch currently:
+
+.. code-block:: python
+    :linenos:
+
+    def _make_sparse(grad, grad_indices, values):
+        size = grad.size()
+        if grad_indices.numel() == 0 or values.numel() == 0:
+            return torch.empty_like(grad)
+        return torch.sparse_coo_tensor(grad_indices, values, size)
+
+    # We don't support sparse gradients
+    param = torch.arange(8).reshape(2, 4).float()
+    state_sum = torch.full_like(param, 0.5)  # initial value for state sum
+
+    grad = grad.coalesce()  # the update is non-linear so indices must be unique
+    grad_indices = grad._indices()
+    grad_values = grad._values()
+    # pow(2) has the same semantics for both sparse and dense memory layouts since 0^2 is zero
+    state_sum.add_(_make_sparse(grad, grad_indices, grad_values.pow(2)))
+
+    # We take care to make std sparse, even though state_sum clearly is not.
+    # This means that we're only applying the gradient to parts of the state_sum
+    # for which it is specified. This even drives the point home a lot more that
+    # the passed gradient is not sparse, but masked.
```

Comment:
Is a pre-requisite of this tutorial knowing what the difference between sparse and masked would be? If not, this definitely requires a bit more explanation. Even the readme of the maskedtensor library https://github.com/pytorch/maskedtensor could be made a lot more effective if it could show a mini example of how a masked tensor is different from a sparse tensor, e.g., with an avg function.

Reply:
I don't think it's particularly a requirement - the main points don't really require the exact semantics differences between the two IMO, but if you disagree please voice!